substrate/tendermint/machine/src/lib.rs

```diff
@@ -0,0 +1,620 @@
+  step_recv: mpsc::UnboundedReceiver<(BlockNumber, Commit<N::SignatureScheme>, N::Block)>,
+
+  block: BlockData<N>,
+}
+
+pub type StepSender<N> = mpsc::UnboundedSender<(
+  BlockNumber,
+  Commit<<N as Network>::SignatureScheme>,
+  <N as Network>::Block,
+)>;
+
+pub type MessageSender<N> = mpsc::UnboundedSender<SignedMessageFor<N>>;
+
+/// A Tendermint machine and its channel to receive messages from the gossip layer over.
+pub struct TendermintHandle<N: Network> {
+  /// Channel to trigger the machine to move to the next height.
+  /// Takes in the the previous block's commit, along with the new proposal.
+  pub step: StepSender<N>,
+  /// Channel to send messages received from the P2P layer.
+  pub messages: MessageSender<N>,
+  /// Tendermint machine to be run on an asynchronous task.
+  pub machine: TendermintMachine<N>,
+}
+
+impl<N: Network + 'static> TendermintMachine<N> {
+  fn broadcast(&mut self, data: DataFor<N>) {
+    if let Some(msg) = self.block.message(data) {
+      self.queue.push_back(msg);
+    }
+  }
+
+  // Start a new round. Returns true if we were the proposer
+  fn round(&mut self, round: RoundNumber, time: Option<CanonicalInstant>) -> bool {
+    if let Some(data) =
+      self.block.new_round(round, self.weights.proposer(self.block.number, round), time)
+    {
+      self.broadcast(data);
+      true
+    } else {
+      false
+    }
+  }
+
+  // 53-54
+  async fn reset(&mut self, end_round: RoundNumber, proposal: N::Block) {
+    // Ensure we have the end time data for the last round
+    self.block.populate_end_time(end_round);
+
+    // Sleep until this round ends
+    let round_end = self.block.end_time[&end_round];
+    sleep(round_end.instant().saturating_duration_since(Instant::now())).await;
+
+    // Clear our outbound message queue
+    self.queue = VecDeque::new();
+
+    // Create the new block
+    self.block = BlockData::new(
+      self.weights.clone(),
+      BlockNumber(self.block.number.0 + 1),
+      self.signer.validator_id().await,
+      proposal,
+    );
+
+    // Start the first round
+    self.round(RoundNumber(0), Some(round_end));
+  }
+
+  async fn reset_by_commit(&mut self, commit: Commit<N::SignatureScheme>, proposal: N::Block) {
+    let mut round = self.block.round().number;
+    // If this commit is for a round we don't have, jump up to it
+    while self.block.end_time[&round].canonical() < commit.end_time {
+      round.0 += 1;
+      self.block.populate_end_time(round);
+    }
+    // If this commit is for a prior round, find it
+    while self.block.end_time[&round].canonical() > commit.end_time {
+      if round.0 == 0 {
+        panic!("commit isn't for this machine's next block");
+      }
+      round.0 -= 1;
+    }
+    debug_assert_eq!(self.block.end_time[&round].canonical(), commit.end_time);
+
+    self.reset(round, proposal).await;
+  }
+
+  async fn slash(&mut self, validator: N::ValidatorId) {
+    if !self.block.slashes.contains(&validator) {
```

Comment:
It's cleared on a new block, as it's contained under new block. It's intended to prevent general spam issues, since this `slash` presumably will cause the validator to make a TX voting to slash at this height. If a malicious validator spams malicious actions, other validators will flood the net.

Reply:
There is potential for multiple slashes in the same block, maybe I am mixing/confusing concepts here and it will not matter, but we do not want to spawn a future bug with this line.

Reply:
Yes, but by doing so enables a DoS since someone triggering a slash causes every other validator to vote on that property, working as Nx multiplier. Slashes also solely penalize the malicious actor. They do not effect algorithmic functionality.